### DISCUSSION TITLE
Open Source Project Contribution

### DISCUSSION THREAD

**Discussion by lexypaul13** [mozilla-mobile/firefox-ios#9524] [General]
Hello guys, I'm an iOS developer with two years of experience. I want to contribute to the development of the firefox mobile app through collaboration and possibly mentorship from skilled iOS engineers. I haven't contributed to an open-source project before; want steps should I take, and what are there prerequisites I must have to contribute?

**Comment by lmarceau** [comment]
Hi @lexypaul13,
Awesome to hear that you want to contribute, you're at the right place! We have some documentation to help you set up a working environment, as well as some explanations on where to find issues you can contribute [here](https://github.com/mozilla-mobile/firefox-ios/wiki/Firefox-iOS-contribution-(Starter-Guide)). My suggestion would be to first have a look there, then ask us any questions if you're unsure of where to start/how. 
Welcome to Firefox iOS, looking forward to your contributions! 😄

**Reply by lexypaul13** [reply]
Hi @lmarceau,
Thank you for the welcome; I cant wait to get started. I'll be sure to reach to you for further questions.

**Reply by lexypaul13** [reply]
Hi @lmarceau,

I hope your day went well. Unfortunately, after cloning the repo on my local machine, an error in the Swifty JSON package module showed when I tried to build the project. Also, the node did not install successfully. The screenshots below show the error message. How do I resolve these issues? 

### Swift Json Error
![Screen Shot 2021-11-19 at 9 08 24 PM](https://user-images.githubusercontent.com/55071531/142710731-8e0533e5-7f31-4ae8-9369-a76e4d6cdebd.png)

### Node Installation Error
![Screen Shot 2021-11-19 at 9 15 08 PM](https://user-images.githubusercontent.com/55071531/142710976-06f3040b-818a-4355-ab65-e9230ee19fca.png)

**Reply by lmarceau** [reply]
Hi @lexypaul13, 
For the project to run, you'll need for sure to have brew carthage and node installed. Did you try to run the following commands?
```
brew update
brew install carthage
brew install node
pip3 install virtualenv
```
Then at the root of the project, run the command `sh ./bootstrap.sh`. It should install all the project requirements, including `SwiftyJSON` through carthage. Only once that's done you can try to run the `Fennec` scheme.

**Reply by lexypaul13** [reply]
Hi @lmarceau,
Hope your day is going well. The installation was not successful. Can you please assist me over a short call via zoom  at your convenient time?

**Reply by lmarceau** [reply]
Hello @lexypaul13,
Could you please let us know which of the steps wasn't successful, and what was the error? If it's an issue with `brew install ...` I would suggest having a look at their forum or asking a question there.

**Reply by lexypaul13** [reply]
@lmarceau When I ran the command sh ./bootstrap.sh in the root file, it showed this error message below. Also, I already have Carthage and virtual env installed, but the installation message for Node states the same thing I shared in the previous screenshot. 


```
A shell task (/usr/bin/xcrun xcodebuild -project /Users/alexpaul/Desktop/firefox-ios/Carthage/Checkouts/A-Star/AStar.xcodeproj CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES -list) failed with exit code 72:
xcrun: error: unable to find utility "xcodebuild", not a developer tool or in PATH. 
```

**Reply by lmarceau** [reply]
@lexypaul13 For Node installation: Make sure you updated brew to the latest with `brew update`. Then you can have a look at what `brew doctor` tells you and go from there? There's certainly a way as well to run the `brew install node` with debug/verbose options to debug what's happening in there, so I would suggest trying that.

For bootstrap, it looks like you need to setup the Command line tools. You can either set it from the command line with `xcode-select` or in Xcode directly. Here's a stackoverflow explaining how to do that in Xcode [here](https://stackoverflow.com/a/41536029/9516941).

**Reply by lexypaul13** [reply]
@lmarceau
 Thanks for the guide on node; I've installed it successfully. However Swifty json isn't installed despite setting up bootstrap on Xcode's location as shown in the image below.
![Screen Shot 2021-11-24 at 10 30 54 PM](https://user-images.githubusercontent.com/55071531/143374989-495d3131-2b4d-4b3f-a7f1-b6855168ca37.png)
d;

**Reply by lmarceau** [reply]
@lexypaul13 
Great to hear node is installed. I guess carthage and virtualenv is installed too correctly? 
Please make sure by running the two following commands in a terminal:
```
brew install carthage
pip3 install virtualenv
```
Once that's done, please navigate to the root of the firefox-ios project with a terminal and run`sh ./bootstrap.sh`. Among other things, this script will install the project dependencies. Only when all those steps are done, then try to run the project with the Fennec target. Note that the screenshot you sent is showing the command line tools, which is something different than Bootstrap.

**Reply by lexypaul13** [reply]
@lmarceau I already installed carthage and virtualenv on my computer. When I ran the command sh ./bootstrap.sh, it showed build complete; however other third party libraries apart from swiftjson didn't install like XCGLogger. See image below
![Screen Shot 2021-11-28 at 10 54 27 PM](https://user-images.githubusercontent.com/55071531/143806319-55e99043-1806-48a5-9682-89ef6a5ef325.png)
![Screen Shot 2021-11-28 at 10 47 57 PM](https://user-images.githubusercontent.com/55071531/143806330-e9e8f257-f6e9-4232-8f8f-3dc2bf6cd12d.png)
.

**Reply by lmarceau** [reply]
Well it looks like something went wrong in the bootstrap installation. If I run it on my side, the completion tells me `5/5` whereas you have `0/0`. Can you see any other messages that would indicate something went wrong in the logs?

**Reply by lmarceau** [reply]
@lexypaul13 Which xcode tools are you using? Are you on Xcode 13.1 or 13.0? If 13.1, I would suggest installing 13.0 tools, seems we're having an issue with 13.1.

**Reply by lexypaul13** [reply]
@lmarceau Hello,  I am using Xcode 13.1. How do I install 13.0 Tools ?

**Reply by lmarceau** [reply]
@lexypaul13 we fixed the 13.1 issue with PR https://github.com/mozilla-mobile/firefox-ios/pull/9587. I would suggest pulling in main, and reinstalling dependencies with `sh ./bootstrap.sh --force`.

**Reply by lexypaul13** [reply]
Hi @lmarceau, after pulling the main, a new error message showed up "Application extensions and any libraries they link to must be built with the `APPLICATION_EXTENSION_API_ONLY` build setting set to YES." Also please have a look at the screenshots below.

![Screen Shot 2021-12-11 at 1 04 29 PM](https://user-images.githubusercontent.com/55071531/145687136-3b4a7f0a-d9c4-4014-8620-f7f1eec570c6.png)

![Screen Shot 2021-12-11 at 1 04 10 PM](https://user-images.githubusercontent.com/55071531/145687137-73627f41-8057-49e2-aa92-ce9fa06d9f54.png)

**Reply by lmarceau** [reply]
@lexypaul13 It seems you have a Carthage issue related to your current `~/.gitconfig`. There's some various thread discussing the issue on the Carthage Github [here](https://github.com/Carthage/Carthage/issues/2671) and [here](https://github.com/Carthage/Carthage/issues/1188). I would suggest having a look at those and try to fix your current setup through that.

**Reply by lexypaul13** [reply]
@lmarceau Hi, I hope your weekend went well. I'll check those resources and other related ones, but I still can't find a solution to fix the Carthage issue. Can we speak via zoom to resolve the problems if it's okay with you?

**Reply by lmarceau** [reply]
Hi @lexypaul13,
We're kind of busy at the moment and wouldn't have time to schedule a zoom meeting, but we can continue to help in the thread. I also suggest that you join the [Matrix group](https://wiki.mozilla.org/Matrix) so you can reach other people there too. The firefox-ios room is `#fx-ios:mozilla.org`

**Reply by lexypaul13** [reply]
Merry Christmas @lmarceau , Ive gotten a new error message shown below. This came right after I cleared the cache in the library folder on my systems. 
Showing Recent Messages
Application extensions and any libraries they link to must be built with the `APPLICATION_EXTENSION_API_ONLY` build setting set to YES.

**Reply by adudenamedruby** [reply]
Hi @lexypaul13. I'm not sure what's going on. Seems like something may have gone awry with your setup? I'd strongly urge you to perhaps delete the `firefox-ios` folder and re-clone from your fork (given that the fork is up to date; if not, please update it). Before doing anything, do make sure you're using the latest `npm` and also run through a full `brew update/upgrade/clean` cycle to make sure everything on your system is up to date. Finally, make sure that you're building using Swift 5.5.1; this means you should be using Xcode 13.1 (13A1030d) as your command line tools.

Then run through the setup steps and let us know what's happening.

**Reply by lexypaul13** [reply]
Hello @electricRGB,
I've tried everything you mentioned but nothing seems to work. I still get the same error message, Could not read serialized diagnostics file: error("Invalid diagnostics signature") (in target 'Shared' from project 'Client')
Command CompileSwift failed with a nonzero exit code. No such module 'SwiftyJSON'
![Screen Shot 2021-12-29 at 12 31 24 PM](https://user-images.githubusercontent.com/55071531/147689619-cdaff607-07b4-48a1-94e5-7c611491cd7b.png)
![Screen Shot 2021-12-29 at 12 23 39 PM](https://user-images.githubusercontent.com/55071531/147689620-e4a3afc3-4594-4e43-b84d-e330972d3663.png)

**Reply by lmarceau** [reply]
@lexypaul13 Seems you still have the same `.gitconfig` issue resulting in `Bad credentials`. Dependencies can't be installed because of that. This isn't an error that relates to our project but rather to your local setup. Please have a look into this and read my previous comment about this [here](https://github.com/mozilla-mobile/firefox-ios/discussions/9524#discussioncomment-1798894).

**Reply by lexypaul13** [reply]
Hello @lmarceau, it been a long time; I hope you've been great. Sorry for not responding, I've been working a full time job. I finally fixed the issue and it runs smoothly. How and where do I start contributing ?

**Reply by lmarceau** [reply]
Hello @lexypaul13, a good place to start is looking into our [contributions guidelines](https://github.com/mozilla-mobile/firefox-ios/blob/main/CONTRIBUTING.md). We also try to mark issues with ["Contributor OK" label](https://github.com/mozilla-mobile/firefox-ios/issues?q=is%3Aissue+is%3Aopen+label%3A%22Contributor+OK%22) to indicate issues that can be a good place to start. If you're unsure about an issue you can ask questions on it. Hope that helps!
